benchmarks/sm3/zscrypto_rv32/sm3.c

```diff
@@ -0,0 +1,122 @@
+
+#include <stdio.h>
+#include <string.h>
+
+#include "riscvcrypto/sm3/api_sm3.h"
+#include "riscvcrypto/share/riscv-crypto-intrinsics.h"
+
+// The block size in bytes
+#define SM3_BLOCK_SIZE (16 * sizeof(uint32_t))
+
+// Reverses the byte order of `V`
+#define REVERSE_BITS_32(V)                                                     \
+  (((V & 0x000000FF) << 24) | (((V)&0x0000FF00) << 8) |                        \
+   (((V)&0x00FF0000) >> 8) | (((V)&0xFF000000) >> 24))
+
+// Rotates `V` by `N` bits to the left
+#define SM3_ROTATE_32(V, N) (((V) << (N)) | ((V) >> (32 - (N))))
+
+// The two permutation functions
+#define SM3_P0(X) _sm3p0((X))
+#define SM3_P1(X) _sm3p1((X))
+
```

Comment:
Note that each expanded word is used only once. Hence you can leverage the large register file and optimize out loads and stores.